### ISSUE TITLE
[Errno 104] Connection reset by peer

### ISSUE THREAD

**Issue by ThePirat91** [sabeechen/hassio-google-drive-backup#351]
###### Description:
```
Hello guys, the following error rose 2 days ago. Until than it had always worked flawlessly. 
Thanks

```

 Addon version: `0.103.1`
 Home Assistant Version: `2021.2.3`
 Supervisor Version: `2021.02.11`
 Supervisor Channel: `stable`
 Hassos Version: `5.11`
 Docker Version: `19.03.13`
 Architecture: `amd64`
 Machine: `qemux86-64`
 Date: `2021-02-19T18:56:50.623088+00:00`
 Timezone: `Europe/Rome`
 Failure Time: `2021-02-19T18:48:38.304281+00:00`
 Last Good Sync: `2021-02-19T18:48:20.392481+00:00`
 ###### Exception:
 ```
 
  addon/backup/model/coordinator.py:156 (_sync)
  addon/backup/model/model.py:180 (sync)
  addon/backup/drive/drivesource.py:160 (save)
  addon/backup/drive/driverequests.py:265 (create)
  addon/backup/drive/driverequests.py:330 (retryRequest)
  addon/backup/creds/driverequester.py:29 (request)
  addon/backup/tracing_session.py:20 (_request)
  /site-packages/aiohttp/client.py:544 (_request)
  /site-packages/aiohttp/client_reqrep.py:890 (start)
  /site-packages/aiohttp/streams.py:604 (read)
aiohttp.client_exceptions.ClientOSError: [Errno 104] Connection reset by peer

 ```
 Snapshots:
 ```
 {
    "HomeAssistant": {
        "snapshots": 2,
        "retained": 0,
        "deletable": 2,
        "name": "HomeAssistant",
        "title": "Home Assistant",
        "latest": "2021-02-19T14:52:51Z",
        "max": 2,
        "enabled": true,
        "size": "268.7 MB",
        "free_space": "17.1 GB"
    },
    "GoogleDrive": {
        "snapshots": 4,
        "retained": 0,
        "deletable": 4,
        "name": "GoogleDrive",
        "title": "Google Drive",
        "latest": "2021-02-16T00:00:07Z",
        "max": 4,
        "enabled": true,
        "size": "857.6 MB"
    }
}
 ```
 ###### Config:
 ```
 {
    "Setting.MAX_SNAPSHOTS_IN_HASSIO": 2,
    "Setting.SNAPSHOT_NAME": "REDACTED",
    "Setting.SNAPSHOT_TIME_OF_DAY": "01:00",
    "Setting.SNAPSHOT_PASSWORD": "REDACTED",
    "Setting.SPECIFY_SNAPSHOT_FOLDER": true,
    "Setting.IGNORE_IPV6_ADDRESSES": true,
    "Setting.GOOGLE_DRIVE_TIMEOUT_SECONDS": 480.0
}
 ```
 ###### Addon Logs:
 ```
 02-19 17:51:37 INFO [backup.model.coordinator] I'll try again in an hour
02-19 18:51:37 DEBUG [backup.model.syncer] Sync requested by Coordinator
02-19 18:51:37 INFO [backup.model.coordinator] Syncing Snapshots
02-19 18:51:37 DEBUG [backup.ha.harequests] Making Hassio request: http://hassio/snapshots
02-19 18:51:37 DEBUG [backup.drive.driverequests] Requesting refreshed Google Drive credentials
02-19 18:51:37 DEBUG [backup.drive.driverequests] Making Google Drive request: https://www.googleapis.com/drive/v3/files/1kbTwEH_559mdX1MlgP7pr5g5Pv4O4AMm/?fields=id%2Cname%2CappProperties%2Csize%2Ctrashed%2CmimeType%2CmodifiedTime%2Ccapabilities%2Cparents&supportsAllDrives=true
02-19 18:51:37 DEBUG [backup.drive.driverequests] Making Google Drive request: https://www.googleapis.com/drive/v3/files/?q=%271kbTwEH_559mdX1MlgP7pr5g5Pv4O4AMm%27+in+parents&fields=nextPageToken%2Cfiles%28id%2Cname%2CappProperties%2Csize%2Ctrashed%2CmimeType%2CmodifiedTime%2Ccapabilities%2Cparents%29&pageSize=100&supportsAllDrives=true&includeItemsFromAllDrives=true&corpora=allDrives
02-19 18:51:38 INFO [backup.drive.drivesource] Uploading 'Meri_2021.2.3_Supervisor v2021.02.11_Full_2021-02-19 15:52:51' to Google Drive
02-19 18:51:38 DEBUG [backup.drive.driverequests] Attempting to resume a previosuly failed upload where we left off
02-19 18:51:38 DEBUG [backup.drive.driverequests] Making Google Drive request: https://www.googleapis.com/upload/drive/v3/files/?uploadType=resumable&supportsAllDrives=true&upload_id=ABg5-UzGgr8xyzQL0M_qa2J4J9iOjtuuAgHFHKVaQFxv14M2atANNgJlc4M4rWNTXrCy9GPPazJZAY78qU9uMLfDoSg-k7V4tQ
02-19 18:59:38 ERROR [backup.model.coordinator] Timed out while trying to reach Google Drive
02-19 18:59:38 INFO [backup.model.coordinator] I'll try again in an hour
02-19 19:48:20 INFO [backup.model.coordinator] Syncing Snapshots
02-19 19:48:20 DEBUG [backup.ha.harequests] Making Hassio request: http://hassio/snapshots
02-19 19:48:20 DEBUG [backup.drive.driverequests] Making Google Drive request: https://www.googleapis.com/drive/v3/files/1kbTwEH_559mdX1MlgP7pr5g5Pv4O4AMm/?fields=id%2Cname%2CappProperties%2Csize%2Ctrashed%2CmimeType%2CmodifiedTime%2Ccapabilities%2Cparents&supportsAllDrives=true
02-19 19:48:20 DEBUG [backup.drive.driverequests] Making Google Drive request: https://www.googleapis.com/drive/v3/files/?q=%271kbTwEH_559mdX1MlgP7pr5g5Pv4O4AMm%27+in+parents&fields=nextPageToken%2Cfiles%28id%2Cname%2CappProperties%2Csize%2Ctrashed%2CmimeType%2CmodifiedTime%2Ccapabilities%2Cparents%29&pageSize=100&supportsAllDrives=true&includeItemsFromAllDrives=true&corpora=allDrives
02-19 19:48:21 INFO [backup.drive.drivesource] Uploading 'Meri_2021.2.3_Supervisor v2021.02.11_Full_2021-02-19 15:52:51' to Google Drive
02-19 19:48:21 DEBUG [backup.drive.driverequests] Attempting to resume a previosuly failed upload where we left off
02-19 19:48:21 DEBUG [backup.drive.driverequests] Making Google Drive request: https://www.googleapis.com/upload/drive/v3/files/?uploadType=resumable&supportsAllDrives=true&upload_id=ABg5-UzGgr8xyzQL0M_qa2J4J9iOjtuuAgHFHKVaQFxv14M2atANNgJlc4M4rWNTXrCy9GPPazJZAY78qU9uMLfDoSg-k7V4tQ
02-19 19:48:21 DEBUG [backup.drive.driverequests] Resuming upload at byte 108527616 of 142223360
02-19 19:48:21 DEBUG [backup.drive.driverequests] Sending 262144 bytes to Google Drive
02-19 19:48:21 DEBUG [backup.drive.driverequests] Making Google Drive request: https://www.googleapis.com/upload/drive/v3/files/?uploadType=resumable&supportsAllDrives=true&upload_id=ABg5-UzGgr8xyzQL0M_qa2J4J9iOjtuuAgHFHKVaQFxv14M2atANNgJlc4M4rWNTXrCy9GPPazJZAY78qU9uMLfDoSg-k7V4tQ
02-19 19:48:21 DEBUG [backup.drive.drivesource] Uploading Meri_2021.2.3_Supervisor v2021.02.11_Full_2021-02-19 15:52:51 76.49%
02-19 19:48:21 DEBUG [backup.drive.driverequests] Sending 4456448 bytes to Google Drive
02-19 19:48:21 DEBUG [backup.drive.driverequests] Making Google Drive request: https://www.googleapis.com/upload/drive/v3/files/?uploadType=resumable&supportsAllDrives=true&upload_id=ABg5-UzGgr8xyzQL0M_qa2J4J9iOjtuuAgHFHKVaQFxv14M2atANNgJlc4M4rWNTXrCy9GPPazJZAY78qU9uMLfDoSg-k7V4tQ
02-19 19:48:34 DEBUG [backup.drive.drivesource] Uploading Meri_2021.2.3_Supervisor v2021.02.11_Full_2021-02-19 15:52:51 79.63%
02-19 19:48:34 DEBUG [backup.drive.driverequests] Sending 3407872 bytes to Google Drive
02-19 19:48:34 DEBUG [backup.drive.driverequests] Making Google Drive request: https://www.googleapis.com/upload/drive/v3/files/?uploadType=resumable&supportsAllDrives=true&upload_id=ABg5-UzGgr8xyzQL0M_qa2J4J9iOjtuuAgHFHKVaQFxv14M2atANNgJlc4M4rWNTXrCy9GPPazJZAY78qU9uMLfDoSg-k7V4tQ
02-19 19:48:38 ERROR [backup.model.coordinator] 
  addon/backup/model/coordinator.py:156 (_sync)
  addon/backup/model/model.py:180 (sync)
  addon/backup/drive/drivesource.py:160 (save)
  addon/backup/drive/driverequests.py:265 (create)
  addon/backup/drive/driverequests.py:330 (retryRequest)
  addon/backup/creds/driverequester.py:29 (request)
  addon/backup/tracing_session.py:20 (_request)
  /site-packages/aiohttp/client.py:544 (_request)
  /site-packages/aiohttp/client_reqrep.py:890 (start)
  /site-packages/aiohttp/streams.py:604 (read)
aiohttp.client_exceptions.ClientOSError: [Errno 104] Connection reset by peer

02-19 19:48:38 INFO [backup.model.coordinator] I'll try again in an hour
 ```
 ###### Supervisor Logs:
 ```
 [32m21-02-19 16:52:39 INFO (MainThread) [supervisor.store] Loading add-ons from store: 126 all - 0 new - 0 remove[0m
[32m21-02-19 16:52:49 INFO (MainThread) [supervisor.homeassistant.api] Updated Home Assistant API token[0m
[32m21-02-19 17:22:49 INFO (MainThread) [supervisor.homeassistant.api] Updated Home Assistant API token[0m
[32m21-02-19 17:51:37 INFO (MainThread) [supervisor.api.security] /snapshots access from cebe7a76_hassio_google_drive_backup[0m
[32m21-02-19 17:51:38 INFO (MainThread) [supervisor.api.security] /snapshots/f9454682/download access from cebe7a76_hassio_google_drive_backup[0m
[32m21-02-19 17:51:38 INFO (MainThread) [supervisor.api.snapshots] Downloading snapshot f9454682[0m
[32m21-02-19 17:52:35 INFO (MainThread) [supervisor.resolution.check] Starting system checks with state CoreState.RUNNING[0m
[32m21-02-19 17:52:35 INFO (MainThread) [supervisor.resolution.check] System checks complete[0m
[32m21-02-19 17:52:36 INFO (MainThread) [supervisor.updater] Fetching update data from https://version.home-assistant.io/stable.json[0m
[32m21-02-19 17:52:51 INFO (MainThread) [supervisor.homeassistant.api] Updated Home Assistant API token[0m
[32m21-02-19 18:05:57 INFO (MainThread) [supervisor.host.info] Updating local host information[0m
[32m21-02-19 18:05:57 INFO (MainThread) [supervisor.host.services] Updating service information[0m
[32m21-02-19 18:05:57 INFO (MainThread) [supervisor.host.network] Updating local network information[0m
[32m21-02-19 18:05:58 INFO (MainThread) [supervisor.host.sound] Updating PulseAudio information[0m
[32m21-02-19 18:05:58 INFO (MainThread) [supervisor.host] Host information reload completed[0m
[32m21-02-19 18:22:53 INFO (MainThread) [supervisor.homeassistant.api] Updated Home Assistant API token[0m
[32m21-02-19 18:48:20 INFO (MainThread) [supervisor.api.security] /snapshots access from cebe7a76_hassio_google_drive_backup[0m
[32m21-02-19 18:48:21 INFO (MainThread) [supervisor.api.security] /snapshots/f9454682/download access from cebe7a76_hassio_google_drive_backup[0m
[32m21-02-19 18:48:21 INFO (MainThread) [supervisor.api.snapshots] Downloading snapshot f9454682[0m
[32m21-02-19 18:48:21 INFO (MainThread) [supervisor.api.security] /snapshots/f9454682/download access from cebe7a76_hassio_google_drive_backup[0m
[32m21-02-19 18:48:21 INFO (MainThread) [supervisor.api.snapshots] Downloading snapshot f9454682[0m
[32m21-02-19 18:49:19 INFO (MainThread) [supervisor.api.security] /supervisor/logs access from cebe7a76_hassio_google_drive_backup[0m
[32m21-02-19 18:49:19 INFO (MainThread) [supervisor.api.security] /core/logs access from cebe7a76_hassio_google_drive_backup[0m
[32m21-02-19 18:51:36 INFO (MainThread) [supervisor.api.security] /supervisor/logs access from cebe7a76_hassio_google_drive_backup[0m
[32m21-02-19 18:51:36 INFO (MainThread) [supervisor.api.security] /core/logs access from cebe7a76_hassio_google_drive_backup[0m
[32m21-02-19 18:52:35 INFO (MainThread) [supervisor.resolution.check] Starting system checks with state CoreState.RUNNING[0m
[32m21-02-19 18:52:35 INFO (MainThread) [supervisor.resolution.check] System checks complete[0m
[32m21-02-19 18:52:55 INFO (MainThread) [supervisor.homeassistant.api] Updated Home Assistant API token[0m
[32m21-02-19 18:56:50 INFO (MainThread) [supervisor.api.security] /supervisor/logs access from cebe7a76_hassio_google_drive_backup[0m

 ```
 ###### Home Assistant Core Logs:
 ```
   hosts = await asyncio.shield(
/usr/local/lib/python3.8/site-packages/alexapy/aiohttp/locks.py:22: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
  self._event = asyncio.Event(loop=loop)
[31m2021-02-19 19:46:43 ERROR (MainThread) [homeassistant.components.xbox] Timeout fetching xbox data[0m
/usr/local/lib/python3.8/site-packages/alexapy/aiohttp/connector.py:999: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
  hosts = await asyncio.shield(
/usr/local/lib/python3.8/site-packages/alexapy/aiohttp/locks.py:22: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
  self._event = asyncio.Event(loop=loop)
[31m2021-02-19 19:47:51 ERROR (MainThread) [homeassistant.components.androidtv.media_player] Failed to execute an ADB command. ADB connection re-establishing attempt in the next update. Error: [Errno 104] Connection reset by peer[0m
[33m2021-02-19 19:48:12 WARNING (MainThread) [homeassistant.components.webhook] Received message for unregistered webhook 9e3ed65f7a8058ddf9409566610d8dc05e2b67e46b0e5b079d3d1913ffebadb4 from 127.0.0.1[0m
/usr/local/lib/python3.8/site-packages/alexapy/aiohttp/connector.py:999: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
  hosts = await asyncio.shield(
/usr/local/lib/python3.8/site-packages/alexapy/aiohttp/locks.py:22: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
  self._event = asyncio.Event(loop=loop)
/usr/local/lib/python3.8/site-packages/alexapy/aiohttp/connector.py:999: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
  hosts = await asyncio.shield(
/usr/local/lib/python3.8/site-packages/alexapy/aiohttp/locks.py:22: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
  self._event = asyncio.Event(loop=loop)
/usr/local/lib/python3.8/site-packages/alexapy/aiohttp/connector.py:999: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
  hosts = await asyncio.shield(
/usr/local/lib/python3.8/site-packages/alexapy/aiohttp/locks.py:22: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
  self._event = asyncio.Event(loop=loop)
[33m2021-02-19 19:51:50 WARNING (MainThread) [homeassistant.components.webhook] Received message for unregistered webhook 9e3ed65f7a8058ddf9409566610d8dc05e2b67e46b0e5b079d3d1913ffebadb4 from 127.0.0.1[0m
[33m2021-02-19 19:51:51 WARNING (MainThread) [homeassistant.components.webhook] Received message for unregistered webhook 9e3ed65f7a8058ddf9409566610d8dc05e2b67e46b0e5b079d3d1913ffebadb4 from 127.0.0.1[0m
/usr/local/lib/python3.8/site-packages/alexapy/aiohttp/connector.py:999: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
  hosts = await asyncio.shield(
/usr/local/lib/python3.8/site-packages/alexapy/aiohttp/locks.py:22: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
  self._event = asyncio.Event(loop=loop)
[31m2021-02-19 19:53:43 ERROR (MainThread) [homeassistant.components.xbox] Timeout fetching xbox data[0m

 ```

**Comment by sabeechen** [comment]
From your logs it looks like the addon is having trouble reaching Google's servers, in just the last 2 hours it got at least one timeout (never got a response form Google) and a connection reset (Google responded initially but gave up the connection after it didn't get anything more).

Connection problems like this are hard to diagnose, because they could be caused by a lot of different things that the addon has no visibility into.  Some causes I've seen users have in the past:

- Recent changes in network hardware (eg a new router).
- Recent failures in network hardware (eg a router thats overheating).
- Spotty internet connection

So far, I've yet to have a bug come in where communication failures like this were caused by a bug in the addon.  The way the addon handles retrys when communication fails is that will keep trying to resume where it left off but with each failure it will bwait a little longer, maxed out at an hour.  After some number of failures (15 IIRC) it will stop trying to resume and just start from the beginning.  This can mean if communication failures are common enough and your snapshot is large, then it may never actually be able to upload the whole snapshot.  I'll be adding some logic in the future to improve this, but for some users connection issue will still always prevent a snapshot form getting uploaded.

**Comment by sabeechen** [comment]
I just released v0.104.1 of the addon, which includes some better handling of flaky connections like this.  If you continue to see problems on the new version please feel free to reopen this bug and let me know.
